### ISSUE TITLE
ReformerLM document representation

### ISSUE THREAD

**Issue by prajdabre** [google/trax#262]
### Description
I have managed to adapt the colab code for learning document representations and the training and generation phase works smoothly. I adapted the sample() method to return the final state after processing the document. However this final state seems to be a complex list consisting of a variety of information. What I want is the hidden representation of the topmost layer which I am assuming represents the whole document. Is there any way to obtain said hidden representation? I am providing my part of the code that is relevant. Any suggestions will be appreciated.
...

### Environment information

```
OS: Ubuntu 16.04 (Irrelevant)

$ pip freeze | grep tensor
mesh-tensorflow==0.1.9
tensor2tensor==1.14.0
tensorboard==1.15.0
tensorflow-datasets==1.3.2
tensorflow-estimator==1.15.1
tensorflow-gan==2.0.0
tensorflow-gpu==1.15.0
tensorflow-hub==0.7.0
tensorflow-metadata==0.21.0
tensorflow-probability==0.7.0


$ pip freeze | grep jax
jax==0.1.57
jaxlib==0.1.37


$ python -V
Python 3.6.8
```

### For bugs: reproduction and error logs

```
# Steps to reproduce:
The relevant part of the code I wrote:

# Prepare a jitted copy of the model.
jit_model_infer = trax.layers.base._accelerate(
    model_infer._forward_internal, trax.math.device_count())
# Set up the initial state for sampling.
infer_state = model_infer.new_weights_and_state(
    trax.supervised.trainer_lib.ShapeDtype((1,1), dtype=np.int32))[1]
infer_state = trainer._for_n_devices(infer_state)

def docvector(length=0, prompt=None):
  """Sample from the ReformerLM model"""
  model_weights = trainer._opt_state[0][0]
  length = len(prompt.split(" "))
  # Token id 0 is the equivalent of a "start" token
  cur_inputs = np.zeros((trax.math.device_count(), 1, 1), dtype=np.int32)

  cur_state = infer_state
  rngs = trax.math.random.split(trax.math.random.get_prng(0), trax.math.device_count())
  all_samples = []

  prompt = np.asarray(
        [TOKENIZER.EncodeAsIds(prompt)] * trax.math.device_count()) <--------- Prompt is the input document as a string.
  logits, cur_state = jit_model_infer(
        cur_inputs,
        model_weights,
        cur_state,
        rngs)
  for iteration in range(length):
    cur_samples = onp.array(prompt[:, iteration], dtype=int)
    cur_inputs = np.array(cur_samples[:,None,None])
    logits, cur_state = jit_model_infer(
        cur_inputs,
        model_weights,
        cur_state,
        rngs)
  
  return cur_state  <-------------------- This is a list of lists/dictionaries/tensors/tuples. How do I get the final hidden state?


# Error logs:
N/A

**Comment by lukaszkaiser** [comment]
The state tree has usually the same form as the model tree, so state[-1] could be the state of the last layer. But the state does not contain activations -- usually (for most layers) it's just empty.

To have the last activations in state, you could just put them there in the `forward_with_state` method of the layer. We just today implemented a caching layer (for decoding) which does that, but only runs the layer once. What you want is something very similar, but simply running and putting the result in the state each time. Take a look at this code and let us know - I'm happy to help more!
https://github.com/google/trax/blob/master/trax/layers/combinators.py#L645

**Comment by prajdabre** [comment]
Hi Lukas,
Thank you for your reply.
I basically understood what you meant. Although I am experienced with TF, I am new to trax and was not able to pinpoint the location where I should be saving states.
Nevertheless, I thought of a quick workaround.
I created a new method called ReformerLMDocRep in which I deleted the dense and logsoftmax from the Map.
Naturally I had to delete the corresponding weights from the optimizer state that is loaded from a pretrained model.
This allowed me to return the output of the last layer of the model as "logits". 
I imagine that this is the representation of the last token (and hence the document vector that I think I am looking for.)
Kindly let me know if you think that my hack is appropriate or not.
Of course if you give me the necessary instructions then I will be able to code up a better solution.
Thanks again for your help as I was able to better understand trax.
